### ISSUE TITLE
Auto-generated proxy doesn't work with endpoints which return actual file content.

### ISSUE THREAD

**Issue by centur** [Azure/autorest#96]
There is an issue with autogenerated proxy - if WebAPI endpoint returns file content something like this:

```c#
public async Task<HttpResponseMessage> DownloadFile(string id)
{
...
    var response = new HttpResponseMessage(HttpStatusCode.OK) { Content = new ByteArrayContent(fileArray) };
    response.Content.Headers.ContentDisposition = new ContentDispositionHeaderValue("attachment") { FileName = itemToDownload.Name };
}
```
And I want to call method on proxy with **WithOperationResponseAsync** suffix which returns  Task<HttpOperationResponse<string>> - assuming that string will be passed as is and not interpret in any way. AutoGenerated proxy reads data as string and tries to deserialize string as json:

``` c#

            string responseContent = await httpResponse.Content.ReadAsStringAsync().ConfigureAwait(false);

...
            // Deserialize Response
            if (statusCode == HttpStatusCode.OK)
            {
                string resultModel = default(string);
                JToken responseDoc = null;
                if (string.IsNullOrEmpty(responseContent) == false)
                {
                    responseDoc = JToken.Parse(responseContent);
                }
                if (responseDoc != null)
                {
                    resultModel = responseDoc.ToString(Newtonsoft.Json.Formatting.Indented);
                }
                result.Body = resultModel;
            }

```

Looks like a bug to me. 
Also it might be good option to support access to raw HttpClient response

**Comment by ogail** [comment]
@centur thanks for reporting this, can you please attach sample Swagger for the repro?

**Comment by centur** [comment]
check this gist [https://gist.github.com/centur/953490b7c5b48c74fc99](https://gist.github.com/centur/953490b7c5b48c74fc99) - I've edited it a bit to remove unrelated bits.
The method in question is "/api/items/{id}" - for Get signature it always generates deserialization logic.

My current fix for this specific item is to edit proxy manually after each proxy generation and add condition to avoid de-serialization logic in case of application/octet-stream response type.

    // Deserialize Response
            if (statusCode == HttpStatusCode.OK)
            {
				if (Equals(httpResponse.Content.Headers.ContentType, new MediaTypeHeaderValue("application/octet-stream")))
				{
					result.Body = responseContent;
				}
				else
				{
					string resultModel = default(string);
					JToken responseDoc = null;
					if (string.IsNullOrEmpty(responseContent) == false)
					{
						responseDoc = JToken.Parse(responseContent);
					}
					if (responseDoc != null)
					{
						resultModel = responseDoc.ToString(Newtonsoft.Json.Formatting.Indented);
					}
					result.Body = resultModel;
				}
			}

**Comment by stankovski** [comment]
@centur thank you for filing this. We were able to repro the issue. As far as the spec goes, however, you should specify the return type to be a file or to have a property of a file and use the appropriate MIME in produces. For example:
```
"produces": [
          "application/octet-stream"
        ],
```
and 
```
  "definitions": {
    "Object": {
      "type": "file"
    },
```

**Comment by centur** [comment]
@stankovski thanks, I've fixed produces section. Can you please elaborate a bit on definitions example? 

If I understood this correctly - I can't just replace ```"type":"file"``` of Object definition (other methods will break as they are  also using Object). 

Can I just set that inside inline response (not sure yet how to do this with Swashbuckle :) but will find out) with  ```"type":"file"``` like this:

``` JSON
 "responses": {
          "200": {
            "description": "OK",
            "schema": {
              "type": "file"
            }
          }
        },

Is this correct ?

**Comment by stankovski** [comment]
Yep, that should work.

**Comment by centur** [comment]
Sorry for the delay: I've got correct swagger definition for the file but it actually got worse :(
We have this definition
``` json
...
"/api/items/{id}": {
      "get": {
        "tags": [
          "Items"
        ],
        "summary": "Download a specific File by Id.",
        "operationId": "Items_DownloadFile",
        "consumes": [],
        "produces": [
          "application/octet-stream"
        ],
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "description": "String ID of the File",
            "required": true,
            "type": "string"
          }
        ],
        "responses": {
          "200": {
            "description": "OK",
            "schema": {
              "type": "file"
            }
          }
        },
        "deprecated": false
      }
  }
...
```

With           "application/octet-stream" in produces section, AutoRest doesn't generate proxy at all

>PS D:\Projects\Tools> .\AutoRest.DocX.ps1
>error: [FATAL] Error generating service model: The given key was not present in the dictionary.
>Finished generating CSharp code for D:\Projects\Tools\Docx.swagger.json.

If I get rid of application/octet-stream and switch it back to 
```
"produces": [
          "application/json",
          "text/json",
          "application/xml",
          "text/xml"
        ],
```
but keep
```
"responses": {
          "200": {
            "description": "OK",
            "schema": {
              "type": "file"
            }
          }
        },
```

Code which is generated by AutoRest is broken (see inline comments for specific errrors):

``` C#
public async Task<HttpOperationResponse<Stream>> DownloadFileWithOperationResponseAsync(string id, CancellationToken cancellationToken = default(System.Threading.CancellationToken))
        {
/*Removed unrelated parts*/
 // Deserialize Response
            if (statusCode == HttpStatusCode.OK)
            {
                Stream resultModel = default(System.IO.Stream);
                JToken responseDoc = null;
                if (string.IsNullOrEmpty(responseContent) == false)
                {
                    responseDoc = JToken.Parse(responseContent);
                }
                if (responseDoc != null)
                {
// Got Error: Cannot implicitly convert type 'string' to 'System.IO.Stream' here
                    resultModel = ((string)responseDoc);
                }
                result.Body = resultModel;
            }
```

So I'm currently switching back to the old fix we had. Hope this will be resolved soon.
AutoRest version is 0.9.7 as latest publicly released one.

**Comment by stankovski** [comment]
Implemented in https://github.com/Azure/autorest/pull/186

**Comment by hovsepm** [comment]
Fixed in Release 0.10.0 - [nuget](https://www.nuget.org/packages/AutoRest/), [chocolatey](https://chocolatey.org/packages/AutoRest), [github] (https://github.com/Azure/autorest/releases/tag/AutoRest-0.10.0)

**Comment by centur** [comment]
Great, thank you
